BUGFIX 4508 mark dependands of live outdated after direct change

mhsdesign opened the pull request:
FIXES: https://github.com/neos/neos-development-collection/issues/4508

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: Test mark dependents of live outdated after direct change

https://github.com/neos/neos-development-collection/issues/4508

mhsdesign committed:
TASK: Improve workspace status behat step and adjust tests to be more precise

mhsdesign committed:
BUGFIX: Calculate workspace state by comparing source cs version and known source version

mhsdesign committed:
TASK: Remove legacy workspace state `UPDATE` logic

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php:
```diff
@@ -754,8 +753,6 @@ private function whenWorkspaceWasCreated(WorkspaceWasCreated $event): void
     private function whenWorkspaceWasDiscarded(WorkspaceWasDiscarded $event): void
     {
         $this->updateWorkspaceContentStreamId($event->workspaceName, $event->newContentStreamId);
-        $this->markWorkspaceAsOutdated($event->workspaceName);
```
this `markWorkspaceAsOutdated` in discard seems to have been wrong. At least i makes no sense because we get a fresh cs.

nezaniel replied:
...and it's gone now anyway ^^

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentRepositoryReadModelAdapter.php:
```diff
@@ -56,11 +55,13 @@ public function findWorkspaceByName(WorkspaceName $workspaceName): ?Workspace
     {
         $workspaceByNameStatement = <<<SQL
             SELECT
```
should we deduplicate this logic using a querybuilder?

bwaidelich replied:
I don't think so

mhsdesign replied:
hmm i did so anyway hope its fine ^^

bwaidelich replied:
Of course it is :)

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentRepositoryReadModelAdapter.php:
```diff
@@ -56,11 +55,13 @@ public function findWorkspaceByName(WorkspaceName $workspaceName): ?Workspace
+                ws.name, ws.baseWorkspaceName, ws.currentContentStreamId, cs.sourceVersion as lastSourceVersion, scs.version as sourceVersion
             FROM
-                {$this->tableNames->workspace()}
+                {$this->tableNames->workspace()} ws
+                JOIN {$this->tableNames->contentStream()} cs ON cs.id = ws.currentcontentstreamid
+                LEFT JOIN {$this->tableNames->contentStream()} scs ON cs.sourceContentStreamId = scs.id
```
so the performance trait is this: we still have the status in the workspace, but we have to calculate it slightly more expensive with these two joins.
I cannot say if this is better performance than having on the write side a similar complex update logic for each event. But as `findWorkspaceByName` (and implicitly  `getContentGraph`) trigger this query and are also part of the write side for constraint checks we have to carefully evaluate.

bwaidelich replied:
At some point the content graph projections needs to know about changes between streams/workspaces (e.g. to count/diff pending changes etc). Then we can revisit this topic and make it faster if needed

mhsdesign commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/Workspaces/WorkspaceState.feature:
```diff
@@ -0,0 +1,109 @@
+      | baseWorkspaceName  | "shared"           |
+      | newContentStreamId | "user-cs-two"      |
+
+  Scenario: Changes to the root workspace render dependents outdated
```
this tests fails on 9.0-dev

bwaidelich replied:
why?

nezaniel replied:
no longer, I guess

mhsdesign committed:
TASK: Rename `sourceContentStreamVersion` column to be more precise

... and refactor status calculation logic. By leveraging `isSatisfiedBy` i found out the `>` comparison was not quite correct. We should use `===`.

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentRepositoryReadModelAdapter.php:
```diff
@@ -136,11 +142,23 @@ public function findContentStreams(): ContentStreams
+        if ($baseWorkspaceName !== null) {
+            // root workspaces can never be out of date and dont have a source content stream
+            $sourceVersion = MaybeVersion::fromVersionOrNull(Version::fromInteger($row['sourceVersion']));
+            $expectedSourceVersion = ExpectedVersion::fromVersion(Version::fromInteger($row['expectedSourceVersion']));
+
+            if (!$expectedSourceVersion->isSatisfiedBy($sourceVersion)) {
+                $status = WorkspaceStatus::OUTDATED;
+            }
```
Mostly a naming thing, but the "expected version" is something for the write side, i.e. I trigger a command and I expect the model to be of a given version.
In this case it is just two versions of content streams.

bwaidelich reviewed (commented):
Looks great already, just a comment regarding "expected version"

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentRepositoryReadModelAdapter.php:
```diff
@@ -136,11 +142,23 @@ public function findContentStreams(): ContentStreams
+            $sourceVersion = MaybeVersion::fromVersionOrNull(Version::fromInteger($row['sourceVersion']));
+            $expectedSourceVersion = ExpectedVersion::fromVersion(Version::fromInteger($row['expectedSourceVersion']));
+
+            if (!$expectedSourceVersion->isSatisfiedBy($sourceVersion)) {
```
rather use equal comparison directly `===`

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentRepositoryReadModelAdapter.php:
```diff
@@ -55,11 +57,13 @@ public function findWorkspaceByName(WorkspaceName $workspaceName): ?Workspace
         $workspaceByNameStatement = <<<SQL
             SELECT
-                name, baseWorkspaceName, currentContentStreamId, status
+                ws.name, ws.baseWorkspaceName, ws.currentContentStreamId, cs.sourceContentStreamVersion as expectedSourceVersion, scs.version as sourceVersion
```
As discussed: Maybe we can simplify this to sth like:
```suggestion
                ws.name, ws.baseWorkspaceName, ws.currentContentStreamId, cs.sourceContentStreamVersion = scs.version as isUpToDate
```

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/4508-mark-dependands-of-live-outdated-after-direct-change

mhsdesign committed:
TASK: Inline variable / select

mhsdesign committed:
TASK: Use `QueryBuilder` to share query

mhsdesign committed:
TASK: Introduce ContentGraphReadModelInterface::getContentGraph to avoid fetching full workspace model

with the extra joins in the workspace model it is just easier to just select the current content stream id directly

mhsdesign committed:
Linte, Linte ging in Laden wollte ein stück käse haben

nezaniel reviewed (approved):
Looks nice

kitsunet commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphReadModelAdapter.php:
```diff
@@ -45,26 +47,44 @@ public function __construct(
-    }
-
-    public function findWorkspaceByName(WorkspaceName $workspaceName): ?Workspace
+    public function getContentGraph(WorkspaceName $workspaceName): ContentGraph
```
not your original sin (actually is I think, just from another change), but like this is the exact problem iwt the find/get thing. This does a DB query every time, so it hsould be more like find? 🤔 

Either way I think we should do something about this, as getContentGraph() is used ALL over the place. This shouldn't be a query everytime.

bwaidelich replied:
good point.. this is insofar a new problem as the previous implementation (of `buildContentGraph()`) did _not_ issue an SQL query..
Not sure what to do about it or if we should ignore the mismatch (for now)

kitsunet replied:
My worry is more the amount of queries we do alone for this call within a single request...

mhsdesign replied:
added todo to readd cache

mhsdesign committed:
Add todo regarding caching

mhsdesign committed:
Add todo regarding exposing and using source content stream version
